src/EventStore.Client/GossipChannelSelector.cs

```diff
@@ -0,0 +1,213 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net;
+using System.Threading;
+using System.Threading.Tasks;
+using Grpc.Core;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Logging.Abstractions;
+
+#nullable enable
+namespace EventStore.Client {
+	internal class GossipChannelSelector : IChannelSelector {
+		private readonly EventStoreClientSettings _settings;
+		private readonly IServerCapabilitiesClient _serverCapabilities;
+		private readonly CancellationToken _cancellationToken;
+
+		private static readonly ClusterMessages.VNodeState[] NotAllowedStates = {
+			ClusterMessages.VNodeState.Manager,
+			ClusterMessages.VNodeState.ShuttingDown,
+			ClusterMessages.VNodeState.Shutdown,
+			ClusterMessages.VNodeState.Unknown,
+			ClusterMessages.VNodeState.Initializing,
+			ClusterMessages.VNodeState.CatchingUp,
+			ClusterMessages.VNodeState.ResigningLeader,
+			ClusterMessages.VNodeState.PreLeader,
+			ClusterMessages.VNodeState.PreReplica,
+			ClusterMessages.VNodeState.Clone,
+			ClusterMessages.VNodeState.DiscoverLeader
+		};
+
+		private readonly Dictionary<EndPoint, ChannelBase> _grpcChannelsCache;
```

Comment:
might be worth a comment explaining what we gain from caching the channels (im not sure i understand it myself actually)

Reply:
Depends on the implementation. In grpc.net it just creates a lot of objects that need to be garbage collected. But in grpc core it ends up calling this: https://github.com/grpc/grpc/blob/master/src/csharp/Grpc.Core/GrpcEnvironment.cs#L117